cmd/snap-confine/udev-support.c

```diff
@@ -18,301 +18,390 @@
+
+	/* Use udev APIs to talk to udev-the-daemon to determine the list of
+	 * "devices" with that tag assigned. The list may be empty, in which case
+	 * there's no udev tagging in effect and we must refrain from constructing
+	 * the cgroup as it would interfere with the execution of a program. */
+	struct udev SC_CLEANUP(sc_cleanup_udev) * udev = NULL;
+	udev = udev_new();
+	if (udev == NULL) {
+		die("cannot connect to udev");
 	}
-	// /dev/nvidia-modeset
-	if (stat(nvidia_modeset_path, &sbuf) == 0) {
-		_run_snappy_app_dev_add_majmin(udev_s, nvidia_modeset_path,
-					       major(sbuf.st_rdev),
-					       minor(sbuf.st_rdev));
+	struct udev_enumerate SC_CLEANUP(sc_cleanup_udev_enumerate) * devices =
+	    NULL;
+	devices = udev_enumerate_new(udev);
+	if (devices == NULL) {
+		die("cannot create udev device enumeration");
 	}
-	// /dev/uhid isn't represented in sysfs, so add it to the device cgroup
-	// if it exists and let AppArmor handle the mediation
-	if (stat("/dev/uhid", &sbuf) == 0) {
-		_run_snappy_app_dev_add_majmin(udev_s, "/dev/uhid",
-					       major(sbuf.st_rdev),
-					       minor(sbuf.st_rdev));
+	if (udev_enumerate_add_match_tag(devices, udev_tag) < 0) {
+		die("cannot add tag match to udev device enumeration");
 	}
-	// When CONFIG_TUN=m, /dev/net/tun will exist but using it doesn't
-	// autoload the tun module but also /dev/net/tun isn't udev tagged
-	// until it is loaded. To work around this, if /dev/net/tun exists, add
-	// it unconditionally to the cgroup and rely on AppArmor to mediate the
-	// access. LP: #1859084
-	if (stat("/dev/net/tun", &sbuf) == 0) {
-		_run_snappy_app_dev_add_majmin(udev_s, "/dev/net/tun",
-					       major(sbuf.st_rdev),
-					       minor(sbuf.st_rdev));
+	if (udev_enumerate_scan_devices(devices) < 0) {
+		die("cannot enumerate udev devices");
 	}
-	// add the assigned devices
-	while (udev_s->assigned != NULL) {
-		const char *path = udev_list_entry_get_name(udev_s->assigned);
-		if (path == NULL)
-			die("udev_list_entry_get_name failed");
-		run_snappy_app_dev_add(udev_s, path);
-		udev_s->assigned = udev_list_entry_get_next(udev_s->assigned);
+	/* NOTE: udev_list_entry is bound to life-cycle of the used udev_enumerate */
+	struct udev_list_entry *assigned;
+	assigned = udev_enumerate_get_list_entry(devices);
+	if (assigned == NULL) {
+		/* NOTE: Nothing is assigned, don't create or use the device cgroup. */
+		debug("no devices tagged with %s, skipping device cgroup setup",
+		      udev_tag);
+		return;
 	}
+
+	sc_cgroup_fds SC_CLEANUP(sc_cleanup_cgroup_fds) fds = { -1, -1, -1 };
+	fds = sc_udev_open_cgroup_v1(security_tag);
+	if (fds.cgroup_procs_fd < 0) {
+		debug("cgroup v1 unavailable, skipping device cgroup setup");
+		return;
+	}
+	/* Setup the device group access control list */
+	sc_udev_setup_acls(fds.devices_allow_fd, fds.devices_deny_fd,
+			   udev, assigned);
+
+	/* Move ourselves to the device cgroup */
+	sc_dprintf(fds.cgroup_procs_fd, "%i", getpid());
```

Comment:
Note, this is a change in behavior. In the old code we moved ourselves into the device cgroup before modifying devices.deny and devices.allow. Is there a reason this isn't before sc_udev_setup_acls()?

Reply:
It felt more correct to move a process only after completing the operations.